### PR TITLE
make deterministic_hash host-architecture-independent

### LINKED ISSUE #38177: assertion failed: eps.windows(2).all(|w| w[0].cmp(self, &w[1]) != Ordering::Greater)
Got an ICE while compiling the `protobuf` crate with that recent snapshot: `rustc 1.15.0-nightly (28d6623bc 2016-12-03)` on arm.

```
thread 'rustc' panicked at 'assertion failed: eps.windows(2).all(|w| w[0].cmp(self, &w[1]) != Ordering::Greater)', /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-host-linux/build/src/librustc/ty/context.rs:1525
```

## Meta

`rustc --version --verbose`:
```
rustc 1.15.0-nightly (28d6623bc 2016-12-03)
binary: rustc
commit-hash: 28d6623bccf8f7e9a40a47895df75dd9ef2619da
commit-date: 2016-12-03
host: arm-unknown-linux-gnueabihf
release: 1.15.0-nightly
LLVM version: 3.9
```

Backtrace: https://gist.github.com/f7ea7c27e5acb684140d2a5227621617

### PR DESCRIPTION
**Pull request by arielb1**
`DefPath::deterministic_hash` used to call `std::hash::Hash`, which depends on the current architecture in several ways, which would prevent metadata written on one host architecture from being successfully read on another one.

Use a hasher we control instead.

Fixes #38177.

r? @michaelwoerister

### CODE REVIEW

**Comment by michaelwoerister** [comment]
I'll take a more detailed look at this tomorrow, but in the meanwhile this might be interesting: https://github.com/rust-lang/rust/blob/aa7a2e9e61cfb9469c7eb88308fa2e1a087ebdb4/src/librustc/ty/util.rs#L398

**Comment by michaelwoerister** [comment]
Thanks for the fix, @arielb1! 

I have few comments though:
- Adding the `StableHash` trait seems a bit ad hoc. We already have something similar [here](https://github.com/rust-lang/rust/blob/01d53df82ef12625f947f5c0a6004e1aea2f9782/src/librustc/session/config.rs#L1680) and it overlaps in functionality with the [ArchIndependentHasher](https://github.com/rust-lang/rust/blob/aa7a2e9e61cfb9469c7eb88308fa2e1a087ebdb4/src/librustc/ty/util.rs#L398)
- You are right about your concerns regarding hashing discriminants, but hashing them explicitly doesn't really scale. We have similar issues with the TypeId hash and also HIR. On the other hand, at least at the moment, the value returned by `discriminant_value` seems to be platform independent. And I don't see any reason why we could not start guaranteeing that it stays that way.

Ideally, we would just use the [IchHasher](https://github.com/rust-lang/rust/blob/01d53df82ef12625f947f5c0a6004e1aea2f9782/src/librustc_incremental/calculate_svh/hasher.rs#L19), which is already built for this purpose (i.e. stable hashing). The module would have to be moved out of `rustc_incremental` though (to `rustc_datastructures` maybe). You can find a 64 bit usage example here: https://github.com/rust-lang/rust/blob/master/src/librustc_trans/partitioning.rs#L214

In the long run we might really want to have a general purpose `StableHash` trait that is supported by a custom derive (with more guarantees for discriminants) and that can be used for type-id hashes, incremental compilation hashes, debuginfo type hashes, and other things, if appropriate.

File: src/libserialize/leb128.rs [hunk 1]
```diff
@@ -9,33 +9,40 @@
...
+        vec[position] = byte;
     }
-
-    *position += 1;
 }
 
-pub fn write_unsigned_leb128(out: &mut Vec<u8>, start_position: usize, mut value: u64) -> usize {
-    let mut position = start_position;
+#[inline]
+pub fn write_unsigned_leb128_to<W>(mut value: u64, mut write: W) -> usize
```
Comment (michaelwoerister) [inline src/libserialize/leb128.rs]: Can you add documentation on what the requirements for `write` are? Something like "`write` is a function that writes a single byte to the output destination at the given position. It is used to write the leb128 value to positions [0..N), where N is the number of bytes the value takes up".

File: src/librustc_data_structures/fingerprint.rs [hunk 2]
```diff
@@ -9,6 +9,7 @@
 // except according to those terms.
```
Comment (michaelwoerister) [inline src/librustc_data_structures/fingerprint.rs]: Can we not move `Fingerprint` to `rustc_datastructures`? It's an incremental compilation specific type.

File: src/librustc_data_structures/stable_hasher.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+    #[inline]
+    fn write(&mut self, bytes: &[u8]) {
+        self.state.write(bytes);
+        self.bytes_hashed += bytes.len() as u64;
+    }
+
+    #[inline]
+    fn write_u8(&mut self, i: u8) {
+        self.state.write_u8(i);
```
Comment (michaelwoerister) [inline src/librustc_data_structures/stable_hasher.rs]: `bytes_hashed` is not incremented here.

File: src/librustc_data_structures/stable_hasher.rs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    #[inline]
+    fn write_usize(&mut self, i: usize) {
+        self.write_uleb128(i as u64);
+    }
+
+    #[inline]
+    fn write_i8(&mut self, i: i8) {
+        self.state.write_i8(i);
```
Comment (michaelwoerister) [inline src/librustc_data_structures/stable_hasher.rs]: Same as above.

File: src/librustc_data_structures/stable_hasher.rs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+///
+/// The hashing just uses the standard `Hash` trait, but the implementations of
+/// `Hash` for the `usize` and `isize` types are *not* architecture independent
+/// (e.g. they has 4 or 8 bytes). As a result we want to avoid `usize` and
+/// `isize` completely when hashing. To ensure that these don't leak in we use a
+/// custom hasher implementation here which inflates the size of these to a `u64`
+/// and `i64`.
+///
+/// The same goes for endianess: We always convert multi-byte integers to little
+/// endian before hashing.
```
Comment (michaelwoerister) [inline src/librustc_data_structures/stable_hasher.rs]: This documentation was appropriate for `ArchIndependentHasher` but the implementation below uses the leb128 approach to solve the problem.

**Review by michaelwoerister — COMMENTED** [review]
Thanks for the updates!

I'm not completely happy with the fact that `StableHasher` is bound to the BLAKE2 algorithm, which is rather heavyweight. But for the time being it makes no difference.
If you address the few nits I'm happy to merge.

**Comment by alexcrichton** [comment]
Nominating for beta to ensure we don't forget about this if it misses the beta branch. Note that this should only go into 1.15.0 beta, not the current (as of this writing) 1.14.0 beta.

**Comment by alexcrichton** [comment]
This made it to beta, de-nominating

**Pull request merged**